crates/compiler/builtins/roc/Json.roc

```diff
@@ -522,123 +486,752 @@ tryDecode = \{ result, rest }, mapper ->
+        (Zero n, b) if b == '.' -> Continue (FractionA (n + 1))
+        (Zero n, b) if isValidEnd b -> Break (Finish n)
+        (Integer n, b) if isDigit0to9 b && n <= maxBytes -> Continue (Integer (n + 1))
+        (Integer n, b) if b == '.' && n < maxBytes -> Continue (FractionA (n + 1))
+        (Integer n, b) if isValidEnd b && n <= maxBytes -> Break (Finish n)
+        (FractionA n, b) if isDigit0to9 b && n <= maxBytes -> Continue (FractionB (n + 1))
+        (FractionB n, b) if isDigit0to9 b && n <= maxBytes -> Continue (FractionB (n + 1))
+        (FractionB n, b) if b == 'e' || b == 'E' && n <= maxBytes -> Continue (ExponentA (n + 1))
+        (FractionB n, b) if isValidEnd b && n <= maxBytes -> Break (Finish n)
+        (ExponentA n, b) if b == '-' || b == '+' && n <= maxBytes -> Continue (ExponentB (n + 1))
+        (ExponentA n, b) if isDigit0to9 b && n <= maxBytes -> Continue (ExponentC (n + 1))
+        (ExponentB n, b) if isDigit0to9 b && n <= maxBytes -> Continue (ExponentC (n + 1))
+        (ExponentC n, b) if isDigit0to9 b && n <= maxBytes -> Continue (ExponentC (n + 1))
+        (ExponentC n, b) if isValidEnd b && n <= maxBytes -> Break (Finish n)
+        _ -> Break Invalid
+
+NumberState : [
+    Start,
+    Minus Nat,
+    Zero Nat,
+    Integer Nat,
+    FractionA Nat,
+    FractionB Nat,
+    ExponentA Nat,
+    ExponentB Nat,
+    ExponentC Nat,
+    Invalid,
+    Finish Nat,
+]
+
+maxBytes : Nat
+maxBytes = 21 # Max bytes in a double precision float
+
+isDigit0to9 : U8 -> Bool
+isDigit0to9 = \b -> b >= '0' && b <= '9'
+
+isDigit1to9 : U8 -> Bool
+isDigit1to9 = \b -> b >= '1' && b <= '9'
+
+isValidEnd : U8 -> Bool
+isValidEnd = \b ->
+    when b is
+        ']' | ',' | ' ' | '\n' | '\r' | '\t' -> Bool.true
+        _ -> Bool.false
 
-            { result: commaResult, rest: nextBytes } = comma beforeCommaOrBreak
+expect
+    actual = "0.0" |> Str.toUtf8 |> Decode.fromBytes fromUtf8
+    expected = Ok 0.0dec
+    actual == expected
 
-            when commaResult is
-                Ok {} -> decodeFields stepField newState nextBytes
-                Err _ -> { result: Ok newState, rest: nextBytes }
+expect
+    actual = "0" |> Str.toUtf8 |> Decode.fromBytes fromUtf8
+    expected = Ok 0u8
+    actual == expected
 
-        { rest: afterBraceBytes } <- bytes |> openBrace |> tryDecode
+expect
+    actual = "1 " |> Str.toUtf8 |> Decode.fromBytesPartial fromUtf8
+    expected = { result: Ok 1dec, rest: [' '] }
+    actual == expected
 
-        { val: endStateResult, rest: beforeClosingBraceBytes } <- decodeFields stepField initialState afterBraceBytes |> tryDecode
+expect
+    actual = "2]" |> Str.toUtf8 |> Decode.fromBytesPartial fromUtf8
+    expected = { result: Ok 2u64, rest: [']'] }
+    actual == expected
 
-        { rest: afterRecordBytes } <- beforeClosingBraceBytes |> closingBrace |> tryDecode
+# TODO why is this trying to decode as a Record?
+# If you use Decode.fromBytesPartial it fails
```

Comment:
@ayazhafiz I'm seeing some strange behaviour here. I think this is being passed to the record decoder for some reason when I use the `Decode.fromBytesPartial`. It is strange because it passes when used with `decodeI64` directly.

Reply:
Runtime error I'm pretty sure. I think its only the assert that is a compiler crash. It shows the failed expectation and variables. But from playing with it I suspect its trying to use the decodeRecord. It wasn't an issue using the old record decoder, and I'm not sure why it would have any relationship with the record decoder.

Reply:
Okay, I will take a look at this today.

Reply:
@ayazhafiz I have been able to narrow this down some more and raised #5252 which I think gets at the root cause of these issues.